src/types.ts

```diff
@@ -2537,6 +2550,38 @@ export class ZodDiscriminatedUnion<
     return this._def.optionsMap;
   }
 
+  _addParentDiscriminator(discriminator: string) {
+    const valueSet = new Set<Primitive>();
+    for (const type of this._def.options) {
+      if (type instanceof ZodObject) {
+        const discriminatorValues = getDiscriminator(type.shape[discriminator]);
+        if (!discriminatorValues.length) {
+          throw new Error(
+            `A discriminator value for key \`${discriminator}\` ` +
+              `could not be extracted from all schema options`
+          );
+        }
+        discriminatorValues.map((el) => valueSet.add(el));
+      } else if (type instanceof ZodDiscriminatedUnion) {
+        const values = type._addParentDiscriminator(discriminator);
+        if (values.length < 1) {
+          throw new Error(
+            `No value for key \`${discriminator}\` was found for DiscriminatedUnion with discriminator \`${type.discriminator}\``
+          );
+        }
+        values.map((el) => valueSet.add(el));
+      }
+    }
+    this._def.parentDiscriminators.add(discriminator);
```

Comment:
What's the reason for adding the parentDiscriminator to `this._def`? It's not used anywhere and also it's a unwanted side-effect. 
Consider ZodDiscriminatedUnion `A` that is added to another ZodDiscriminatedUnion `B`. Now when `A` is used directly it wrongfully has `B` as parent discriminator. Also when later added to `C` it now has 2 parent discriminators.

Reply:
Hey @teunmooij thanks for your feedback! I think this was leftover when I was thinking of how to solve this earlier (pre disjointed discriminators). For some reason I can't recall I thought it was useful; but yeah no need for this at the moment. The method can just recurse without having to store the parent discriminators around.

Reply:
resolved with [223b34b](https://github.com/colinhacks/zod/pull/1589/commits/223be4b0cc835b3352ed795335d6ff8f55f2d3ed)